### DISCUSSION TITLE
README: "AST doesn't know about spaces"

### DISCUSSION THREAD

**Discussion by mindplay-dk** [rectorphp/rector#8538] [General]
is this still pertinent?

https://github.com/rectorphp/rector/blob/3e8c19254ee9017d210d859ecd73fd3733ecc867/README.md?plain=1#L152

according to [this issue](https://github.com/nikic/PHP-Parser/issues/344), most issues with spacing were long since resolved?

in my preliminary testing with Rector, I have not seen it mangle any spaces - and that's in a codebase with no code standard 🙂

I understand there may still be some issues with spacing, but I almost skipped Rector (and would have done all of my rename operations by hand in PHP Storm instead) because of this warning - and clearly, it goes have some level of awareness of spacing, or this project would most likely have been totally mangled. 😅

(glad I chose Rector, as this will save me a *lot* of time! thank you for this incredible tool!)

**Comment by TomasVotruba** [comment]
Hi, thanks for the question. You're right :slightly_smiling_face: , it can be serious warning or not relevant at all, it depends on your codebase.

We've had many opposite reactions, where people wanted all the spaced to remain the same, but technically it was not possible, so we added this warning to expect the edge-case issues (depending on your code base) and be prepated.

**Reply by mindplay-dk** [reply]
my point was this was probably written at a time when truly the AST did "now know about spaces" - of course, when you refactor code automatically, it affects spaces too in some cases, but the way this was worded, I basically expected the tool to reformat my entire codebase, which, thankfully, is not the case 😅

**Reply by TomasVotruba** [reply]
It's better now, but still relevant :D

**Comment by samsonasik** [comment]
@mindplay-dk when no change on the target file, there is tweak on rector when comparing output from php-parser printer vs original file content:

https://github.com/rectorphp/rector/blob/3e8c19254ee9017d210d859ecd73fd3733ecc867/src/Application/FileProcessor.php#L161-L181

with php-parser v5, the printed content no longer `ltrim`-ed, which we will revisit later once phpstan go to v2 which will require php-parser v5

**Comment by TomasVotruba** [comment]
Converted to discussion, as it fits better :+1:
